core/modules/replica/Job.h

```diff
@@ -0,0 +1,482 @@
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+// Forward declarations
+namespace database {
+namespace mysql {
+class Connection;
+}}
+
+/**
+  * Class Job is a base class for a family of replication jobs within
+  * the master server.
+  */
+class Job
+    :   public std::enable_shared_from_this<Job>  {
+
+public:
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<Job> Ptr;
+
+    /// The pointer type for the database connector which provides a database-specific
+    /// SQL generation services.
+    typedef std::shared_ptr<database::mysql::Connection> SqlGeneratorPtr;
+
+    /// Primary public state of the job
+    enum State {
+
+        /// The job has been constructed, and no attempt to execute it has
+        /// been made.
+        CREATED,
+
+        /// The job is in a progress
+        IN_PROGRESS,
+
+        /// The job is finihed. See extended status for more details
+        /// (the completion status, etc.)
+        FINISHED
+    };
+
+    /// Return the string representation of the primary state
+    static std::string state2string(State state) ;
+
+    /// Refined public sub-state of the job once it's FINISHED as per
+    /// the above defined primary state.
+    enum ExtendedState {
+
+        /// No extended state exists at this time
+        NONE,
+
+        /// The job has been fully implemented
+        SUCCESS,
+
+        /// Problems with job configuration found
+        CONFIG_ERROR,
+
+        /// The job has failed
+        FAILED,
+
+        /// Qserv notification failed
+        QSERV_FAILED,
+
+        /// Qserv reported that the source chunk is in use and couldn't be removed
+        QSERV_IN_USE,
+
+        /// Expired due to a timeout (as per the Configuration)
+        EXPIRED,
+
+        /// Explicitly cancelled on the client-side (similar to EXPIRED)
+        CANCELLED
+    };
+
+    /// @return string representation of the extended state
+    static std::string state2string(ExtendedState state) ;
+
+    /// @return string representation of the combined state
+    static std::string state2string(State state, ExtendedState extendedState) {
+        return state2string(state) + "::" +state2string(extendedState);
+    }
+
+    /// The job options container
+    struct Options {
+
+        /// the priority level
+        int  priority;
+
+        /// the flag indicating of this job can't be run simultaneously
+        /// along with other jobs.
+        bool exclusive;
+
+        /// the flag indicating if the job is allowed to be interrupted by some
+        /// by other jobs.
+        bool preemptable;
+    };
+
+    // Default construction and copy semantics are prohibited
+
+    Job() = delete;
+    Job(Job const&) = delete;
+    Job& operator=(Job const&) = delete;
+
+    virtual ~Job() = default;
+
+    /// @return a reference to the Controller,
+    Controller::Ptr controller() const { return _controller; }
```

Comment:
I forget if the return Ptr is supposed to be `const &`?

Reply:
Fixed